0000-modules/detailed-design/loading-files.md

```diff
@@ -0,0 +1,231 @@
+# Loading Files
+
+When building a Rust project, rustc will load and parse some files as Rust code
+in addition to the root module. These will be used to construct a module tree.
+By default, cargo will generate the list of files to load in this way for you,
+though you can generate such a list yourself and specify it in your
+`Cargo.toml`, or you can generate the list in another way for your non-cargo
+build system.
+
+This eliminates the need to write `mod` statements to add new files to your
+project. Instead, files will be picked up automatically as a part of your
+module hierarchy.
+
+## Detailed design
+
+### Processing the `--modules` list (rustc)
+
+rustc takes a new argument called `modules`, which takes a space separated list
+of files. Each file will be treated as a module, and rustc will attempt to open
+and parse every file listed, reporting errors if it is unable to. It will mount
+these files as a tree of Rust modules using rules which mirror the current
+rules for looking up modules.
+
+It will not attempt to open or parse files where:
+
+* The file name is not a valid Rust identifier followed by `.rs`.
+* The file is not in a subdirectory of the directory containing the root
+module.
```

Comment:
Depending on how this "subdirectory of" check is implemented it could get pretty tricky:

* Do the prefixes given to `--modules` have to match? If not:
* How are paths normalised? How are soft/hard links handled?

In most cases this probably does not matter, but it may still be worth considering.